aggregator/src/aggregator/aggregation_job_driver.rs

```diff
@@ -740,6 +780,91 @@ where
             aggregation_job,
             stepped_aggregations,
             report_aggregations_to_write,
+            retry_after.as_ref(),
+            resp,
+        )
+        .await
+    }
+
+    async fn step_aggregation_job_aggregate_poll<
+        const SEED_SIZE: usize,
+        C: Clock,
+        B: CollectableBatchMode,
+        A: vdaf::Aggregator<SEED_SIZE, 16> + Send + Sync + 'static,
+    >(
+        &self,
+        datastore: Arc<Datastore<C>>,
+        vdaf: Arc<A>,
+        lease: Arc<Lease<AcquiredAggregationJob>>,
+        task: Arc<AggregatorTask>,
+        aggregation_job: AggregationJob<SEED_SIZE, B, A>,
+        report_aggregations: Vec<ReportAggregation<SEED_SIZE, A>>,
+    ) -> Result<(), Error>
+    where
+        A::AggregationParam: Send + Sync + PartialEq + Eq,
+        A::AggregateShare: Send + Sync,
+        A::InputShare: Send + Sync,
+        A::OutputShare: Send + Sync,
+        A::PrepareState: Send + Sync + Encode,
+        A::PrepareShare: Send + Sync,
+        A::PrepareMessage: Send + Sync,
+        A::PublicShare: Send + Sync,
+    {
+        // Only process non-failed report aggregations; convert non-failed report aggregations into
+        // stepped aggregations to be compatible with `process_response_from_helper`.
+        let stepped_aggregations: Vec<_> = report_aggregations
+            .into_iter()
+            .filter_map(|report_aggregation| {
+                let leader_state = match report_aggregation.state() {
+                    ReportAggregationState::LeaderPoll { leader_state } => {
+                        Some(leader_state.clone())
+                    }
+                    _ => None,
+                };
+
+                leader_state.map(|leader_state| SteppedAggregation {
+                    report_aggregation,
+                    leader_state,
+                })
```

Comment:
This is a potential candidate for future optimizations to reduce peak memory usage. Two copies of the leader state could be expensive for large instantiations of Prio3. Maybe we could put this in an Arc, or change SteppedAggregation to handle jobs in the poll state without this duplication.

Reply:
Agreed -- using an `Arc`, or just moving the large value (where possible), would improve things here with little tradeoff.